osf/metrics/reporters/institutional_users.py

```diff
@@ -0,0 +1,135 @@
+import dataclasses
+import datetime
+
+from django.contrib.contenttypes.models import ContentType
+from django.db.models import Q, F, Sum
+
+from osf import models as osfdb
+from osf.models.spam import SpamStatus
+from addons.osfstorage.models import OsfStorageFile
+from osf.metrics.reports import InstitutionalUserReport
+from osf.metrics.utils import YearMonth
+from ._base import MonthlyReporter
+
+
+_CHUNK_SIZE = 500
+
+
+class InstitutionalUsersReporter(MonthlyReporter):
+    def report(self, yearmonth: YearMonth):
+        _before_datetime = yearmonth.next_month()
+        for _institution in osfdb.Institution.objects.all():
+            _user_qs = _institution.get_institution_users().filter(created__lt=_before_datetime)
+            for _user in _user_qs.iterator(chunk_size=_CHUNK_SIZE):
+                _helper = _InstiUserReportHelper(_institution, _user, yearmonth, _before_datetime)
+                yield _helper.report
+
+
+# helper
+@dataclasses.dataclass
+class _InstiUserReportHelper:
+    institution: osfdb.Institution
+    user: osfdb.OSFUser
+    yearmonth: YearMonth
+    before_datetime: datetime.datetime
+    report: InstitutionalUserReport = dataclasses.field(init=False)
+
+    def __post_init__(self):
+        _affiliation = self.user.get_institution_affiliation(self.institution._id)
+        self.report = InstitutionalUserReport(
+            report_yearmonth=self.yearmonth,
+            institution_id=self.institution._id,
+            user_id=self.user._id,
+            user_name=self.user.fullname,
+            department_name=(_affiliation.sso_department or None),
+            month_last_login=YearMonth.from_date(self.user.date_last_login),
+            account_creation_date=YearMonth.from_date(self.user.created),
+            orcid_id=self.user.get_verified_external_id('ORCID', verified_only=True),
+            public_project_count=self._public_project_queryset().count(),
+            private_project_count=self._private_project_queryset().count(),
+            public_registration_count=self._public_registration_queryset().count(),
+            embargoed_registration_count=self._embargoed_registration_queryset().count(),
+            public_file_count=self._public_osfstorage_file_queryset().count(),
+            published_preprint_count=self._published_preprint_queryset().count(),
+            storage_byte_count=self._storage_byte_count(),
+        )
+
+    def _node_queryset(self):
+        _institution_node_qs = self.institution.nodes.filter(
+            created__lt=self.before_datetime,
+            is_deleted=False,
+        ).exclude(spam_status=SpamStatus.SPAM)
+        return osfdb.Node.objects.get_nodes_for_user(
+            user=self.user,
+            base_queryset=_institution_node_qs,
+        )
+
+    def _public_project_queryset(self):
+        return self._node_queryset().filter(
+            type='osf.node',  # `type` field from TypedModel
```

Comment:
nit: Good comments for each function, but it would be nice to have a docstring on the class listing these out.